contentcuration/contentcuration/assets/shared/views/ThumbnailUpload.vue

```diff
@@ -0,0 +1,135 @@
+<template>
+  <div
+    :style="highlight && !readonly ? { borderColor: 'purple' } : {}"
+    @dragenter.prevent="enter"
+    @dragover.prevent="over"
+    @dragleave.prevent="leave"
+    @drop.prevent="drop"
+  >
+    <template
+      v-for="file in files"
+    >
+      <VProgressCircular
+        v-if="file.progress < 100"
+        :key="file.id"
+        :size="50"
+        :width="15"
+        :value="file.progress"
+      >
+        {{ file.progress }}
+      </VProgressCircular>
+    </template>
+    <input
+      v-if="!readonly"
+      id="fileElem"
+      ref="fileUpload"
+      style="display: none;"
+      type="file"
+      accept="image/*"
+      @change="handleFiles($event.target.files)"
+    >
+    <label for="fileElem">
+      <VImg v-if="file" contain :src="file" />
+      <VImg v-else contain :src="require('../images/kolibri-logo.svg')" />
+      <VBtn v-if="!readonly" @click="$refs.fileUpload.click()">
+        {{ $tr('fileUpload') }}
+      </VBtn>
+    </label>
+  </div>
+</template>
+
+<script>
+
+  import client from 'shared/client';
+
+  export default {
+    name: 'ThumbnailUpload',
+    $trs: {
+      fileUpload: 'Upload file',
+    },
+    props: {
+      value: {
+        type: Object,
+        default() {
+          return {};
+        },
+      },
+      readonly: {
+        type: Boolean,
+        default: false,
+      },
+    },
+    data() {
+      return {
+        highlight: false,
+        files: [],
+      };
+    },
+    computed: {
+      file() {
+        if (this.files.length && this.files[0].previewSrc) {
+          return this.files[0].previewSrc;
+        }
+        return this.value.thumbnail_url;
+      },
+    },
+    methods: {
+      enter() {
+        this.highlight = true;
+      },
+      over() {
+        this.highlight = true;
+      },
+      leave() {
+        this.highlight = false;
+      },
+      drop(e) {
+        this.highlight = false;
+        this.handleFiles(e.dataTransfer.files);
+      },
+      handleFiles(files) {
+        if (!readonly) {
+          [...files].forEach(file => {
+            const id = String(Math.random()).slice(2);
+            const fileDetails = {
+              id,
+              previewSrc: null,
+              progress: null,
+            };
+            this.files.push(fileDetails);
+            const data = new FormData();
+            data.append('file', file);
+
+            client
+              .post(window.Urls.thumbnail_upload(), data, {
+                headers: {
+                  'Content-Type': 'multipart/form-data',
+                },
+                onUploadProgress: progressEvent => {
+                  const { loaded, total } = progressEvent;
+                  fileDetails.progress = (loaded / total) * 100;
+                },
+              })
+              .then(response => {
+                this.$emit('input', {
```

Comment:
Once it has finished - emit an `input` event to fit into the v-model flow - the component embedding this one is responsible for updating state with the output.

Reply:
perhaps a good candidate for https://vuejs.org/v2/guide/components-custom-events.html#sync-Modifier

Reply:
Hrm - I think I prefer the `v-model` flow myself - because it aligns with how input components are handled, but open to other POV.

Reply:
I don't have a strong argument for one way or the other